tiberius/src/lib.rs

```diff
@@ -640,44 +641,68 @@ impl<'a> ToConnectEndpoint<Box<BoxableIo>, DynamicConnectionTarget> for &'a str
-                Some(key_end) => ((&input[..key_end]).to_lowercase(), key_end)
+            // (MSDN) The basic format of a connection string includes a series of keyword/value
+            // pairs separated by semicolons. The equal sign (=) connects each keyword and its value.
+            let key = {
+                let mut t = input.splitn(2, '=');
+                let key = t.next().unwrap();
+                if let Some(i) = t.next() {
+                    input = i;
+                } else {
+                    return Err(TdsError::Conversion("connection string expected key. expected `=` never found".into()))
+                }
+                key.trim().to_lowercase()
             };
-            input = &input[key_end+1..];
-            // check if an escaped value (e.g. 'my password contains a;' or 'or a \' #kappa' follows)
-            let escaped = input.starts_with('\'');
-            let end = input.bytes().position(|x| x == b';').unwrap_or(input.len());
-
-            let value = if !escaped {
-                let ret = &input[..end];
-                input = &input[cmp::min(end + 1, input.len())..];
-                Cow::Borrowed(ret)
-            } else {
-                let mut val = String::with_capacity(end);
+
+            // (MSDN) To include values that contain a semicolon, single-quote character,
+            // or double-quote character, the value must be enclosed in double quotation marks.
+            // If the value contains both a semicolon and a double-quote character, the value can
+            // be enclosed in single quotation marks.
+            let quote_char = match &input[0..1] {
+                "'" => Some('\''),
+                "\"" => Some('"'),
+                _ => None,
+            };
+
+            let value = if let Some(quote_char) = quote_char {
+                let mut value = String::new();
                 loop {
-                    let next = match input.bytes().position(|x| x == b'\'' || x == b'\\') {
-                        None => return Err(TdsError::Conversion("connection string: unterminated escape sequence".into())),
-                        Some(next) => next
-                    };
-                    match input.as_bytes()[next] {
-                        b'\'' => {
-                            val.push_str(&input[..next]);
-                            break;
-                        },
-                        b'\\' if next+1 < input.len() && input.as_bytes()[next+1] == b'\'' => {
-                            val.push('\'');
-                            input = &input[1..]; // <=> &[next+2..] effectively
-                        },
-                        b'\\' => val.push('\\'),
-                        _ => unreachable!()
+                    let mut t = input.splitn(3, quote_char);
+                    t.next().unwrap(); // skip first quote character
+                    value.push_str(t.next().unwrap());
```

Comment:
This might also panic, if the input only consists of a single quote character. Don't panic here.

Reply:
Same here. The first character must be `quote_char` regardless of the input (guaranteed by the `if`s), then at least two first iterator values are always valid. If it panics here, it means a bug somewhere in `if` statements below. There is no user input string that could cause failure here.

Reply:
An example demonstrating how it works:
```
let mut t = "".splitn(2, ",");                                                                  
t.next().unwrap();                                                                              
assert_eq!(t.next(), None);                                                                                                                                                      
                                                                                             
let mut t = ",".splitn(3, ",");                                                                 
t.next().unwrap();                                                                              
t.next().unwrap();                                                                              
assert_eq!(t.next(), None);                                                                     
```

Reply:
Okay, wasn't clear to me from the docs, thanks.
I don't quite the unwraps (since it isn't immediately obvious that those are "safe"),
but atleast it works.

Reply:
In our company we tend to write `.expect("Bug: expected something blablabla")` in such places, so that, if a crash is reported, the developer sees the cause immediately.

Thank you!